data_structures/avltree_map.cpp

```diff
@@ -0,0 +1,226 @@
+}
+
+void testClear()
+{
+    TestMap map = initWithKeys({ 3, 7, 5, 0, 9, 4, 6, 2, 1, 8 });
+    map.clear();
+    ASSERT_EQUAL(map.empty(), true);
+}
+
+void testGetAndRemove()
+{
+    TestMap map = initWithKeys({ 3, 7, 5, 0, 9, 4, 6, 2, 1, 8 });
+    std::vector<int> removes = { 7, 0, 3, 5, 1 };
+    for (const auto & item: removes) {
+        ASSERT_EQUAL(map.getAndRemove(item), item);
+    }
+    auto result = AvlTreeMap<int, int>::EntryList();
+    ASSERT_EQUAL(map.toEntryList(), TestMap::EntryList({
+        {2, 2}, {4, 4}, {6, 6}, {8, 8}, {9, 9}
+    }));
+}
+
+void testRemoveAll()
+{
+    TestMap map = initWithKeys({ 3, 7, 5, 0, 9, 4, 6, 2, 1, 8 });
+    map.removeAll([](int k, int v) -> bool {
+        return k % 2 != 0;
+    });
+    auto result = AvlTreeMap<int, int>::EntryList();
+    ASSERT_EQUAL(map.toEntryList(), TestMap::EntryList({
+        {0, 0}, {2, 2}, {4, 4}, {6, 6}, {8, 8}
+    }));
+}
+
+void testGetCeilingEntry()
+{
+    TestMap map = initWithKeys({ 1, 5, 9, 4, 0, 6 });
+    ASSERT_EQUAL(map.getCeilingEntry(5).key, 5);
+    ASSERT_EQUAL(map.getCeilingEntry(6).key, 6);
+    ASSERT_EQUAL(map.getCeilingEntry(2).key, 4);
+    ASSERT_EQUAL(map.getCeilingEntry(7).key, 9);
+    ASSERT_EQUAL(map.getCeilingEntry(-1).key, 0);
+    ASSERT_THROW(map.getCeilingEntry(100), TestMap::NoSuchMappingException);
+}
+
+void testGetFloorEntry()
+{
+    TestMap map = initWithKeys({ 1, 5, 9, 4, 0, 6 });
+    ASSERT_EQUAL(map.getFloorEntry(5).key, 5);
+    ASSERT_EQUAL(map.getFloorEntry(6).key, 6);
+    ASSERT_EQUAL(map.getFloorEntry(2).key, 1);
+    ASSERT_EQUAL(map.getFloorEntry(7).key, 6);
+    ASSERT_EQUAL(map.getFloorEntry(100).key, 9);
+    ASSERT_THROW(map.getFloorEntry(-1), TestMap::NoSuchMappingException);
+}
+
+void testGetHigherEntry()
+{
+    TestMap map = initWithKeys({ 1, 5, 9, 4, 0, 6 });
+    ASSERT_EQUAL(map.getHigherEntry(5).key, 6);
+    ASSERT_EQUAL(map.getHigherEntry(6).key, 9);
+    ASSERT_EQUAL(map.getHigherEntry(2).key, 4);
+    ASSERT_EQUAL(map.getHigherEntry(7).key, 9);
+    ASSERT_EQUAL(map.getHigherEntry(-1).key, 0);
+    ASSERT_THROW(map.getHigherEntry(100), TestMap::NoSuchMappingException);
+}
+
+void testGetLowerEntry()
+{
+    TestMap map = initWithKeys({ 1, 5, 9, 4, 0, 6 });
+    ASSERT_EQUAL(map.getLowerEntry(5).key, 4);
+    ASSERT_EQUAL(map.getLowerEntry(6).key, 5);
+    ASSERT_EQUAL(map.getLowerEntry(2).key, 1);
+    ASSERT_EQUAL(map.getLowerEntry(7).key, 6);
+    ASSERT_EQUAL(map.getLowerEntry(100).key, 9);
+    ASSERT_THROW(map.getLowerEntry(-1), TestMap::NoSuchMappingException);
+}
+
+int main()
+{
+    RUN_TEST(testInsert);
+    RUN_TEST(testAssign);
+    RUN_TEST(testRemove);
+    RUN_TEST(testGet);
+    RUN_TEST(testSize);
+    RUN_TEST(testClear);
+    RUN_TEST(testGetOrInsert);
+    RUN_TEST(testGetAndRemove);
+    RUN_TEST(testRemoveAll);
+    RUN_TEST(testGetCeilingEntry);
+    RUN_TEST(testGetFloorEntry);
+    RUN_TEST(testGetHigherEntry);
+    RUN_TEST(testGetLowerEntry);
+}
```

Comment:
Missing an endline. Without this endline, there could be issues with some IDEs and configurations.
```suggestion
}

```